ä¸‹é¢æˆ‘æŠŠä½ æä¾›çš„è„šæœ¬æŒ‰ **æ¯ä¸€è¡Œï¼ˆæˆ–æ¯ä¸€è¯­å¥ï¼‰** ç”¨ä¸­æ–‡åŠ å…¥å°½å¯èƒ½è¯¦ç»†çš„æ³¨é‡Šï¼Œå¹¶åœ¨æœ€åç»™å‡ºä»£ç åŠŸèƒ½çš„æ€»ç»“ä¸å…³é”®æ³¨æ„ç‚¹ã€‚å› ä¸ºåŸå§‹è„šæœ¬æ¯”è¾ƒé•¿ï¼Œæˆ‘å¯¹æ¯ä¸€è¡Œéƒ½å†™äº†è¯´æ˜ï¼ˆå¯¹å¾ˆç±»ä¼¼çš„é‡å¤è¡Œä¼šä¿æŒæ³¨é‡Šé£æ ¼ä¸€è‡´ä»¥æé«˜å¯è¯»æ€§ï¼‰ï¼Œå¦‚æœä½ å¸Œæœ›æˆ‘æŠŠæ³¨é‡Šç›´æ¥å†™å›æºæ–‡ä»¶ä¸­ï¼ˆå³ç”Ÿæˆä¸€ä¸ªå¸¦æ³¨é‡Šçš„æ–° `.py` æ–‡ä»¶ï¼‰ï¼Œå‘Šè¯‰æˆ‘æˆ‘å¯ä»¥æŠŠå®Œæ•´æ³¨é‡Šç‰ˆæœ¬ä¿å­˜ä¸ºæ–‡ä»¶å¹¶ç»™å‡ºä¸‹è½½é“¾æ¥ã€‚

> è¯´æ˜ï¼šæ³¨é‡Šä»¥ `# ` å¼€å¤´ï¼Œå¹¶æ”¾åœ¨å¯¹åº”ä»£ç è¡Œä¹‹åæˆ–ä¸Šä¸€è¡Œã€‚ä¸ºæ–¹ä¾¿é˜…è¯»æˆ‘ä¿æŒä»£ç åŸæ ·å¹¶åœ¨æ¯è¡Œåæˆ–ä¸Šä¸€è¡ŒåŠ æ³¨é‡Šã€‚

---

```python
"""Cong Liu Yi_lab, Peking University"""
"""Feb, 2021"""
"""Email: liucong-1112@pku.edu.cn"""
"""Usage: This program is used to transfor sites to genome locus"""
"""Input: [.bam]"""
# â€”â€”â€” æ–‡ä»¶å¤´æ³¨é‡Šï¼šä½œè€…ã€æ—¥æœŸã€è”ç³»é‚®ç®±ã€ç”¨é€”ã€è¾“å…¥ç±»å‹è¯´æ˜ï¼ˆåŸæ–‡ï¼‰

import os, sys, time, argparse, pysam
# å¯¼å…¥æ ‡å‡†åº“å’Œç¬¬ä¸‰æ–¹åº“ï¼š
# os: æ–‡ä»¶/è·¯å¾„/ç³»ç»Ÿç›¸å…³æ“ä½œ
# sys: ä¸ Python è§£é‡Šå™¨äº¤äº’ï¼ˆä¾‹å¦‚é€€å‡ºã€æ ‡å‡†æµï¼‰
# time: æ—¶é—´ç›¸å…³å‡½æ•°ï¼ˆä¾‹å¦‚ localtimeï¼‰
# argparse: å‘½ä»¤è¡Œå‚æ•°è§£æ
# pysam: å¤„ç† BAM/SAM æ–‡ä»¶ï¼ˆåŸºäº htslib çš„ Python ç»‘å®šï¼‰
from Bio import SeqIO
# Biopython çš„ SeqIO ç”¨äºè¯»å– FASTA ç­‰åºåˆ—æ–‡ä»¶
from Bio.Seq import reverse_complement
# Biopython çš„ reverse_complementï¼Œç”¨äºå¯¹åºåˆ—æŒ‰ç¢±åŸºäº’è¡¥å¹¶åå‘
from time import strftime
# ä» time ä¸­å¯¼å…¥ strftimeï¼Œç”¨äºæ ¼å¼åŒ–æ—¶é—´è¾“å‡º
from collections import defaultdict
# defaultdict æä¾›å¸¦é»˜è®¤å€¼çš„å­—å…¸ï¼ˆå½“é”®ä¸å­˜åœ¨æ—¶è‡ªåŠ¨åˆ›å»ºï¼‰
from collections import OrderedDict
# OrderedDict ä¿æŒæ’å…¥é¡ºåºçš„å­—å…¸ï¼ˆåœ¨éƒ¨åˆ† Python ç‰ˆæœ¬ä¸­æœ‰æ„ä¹‰ï¼‰
from pysam import qualities_to_qualitystring
# pysam ä¸­å°†è´¨é‡å€¼åˆ—è¡¨è½¬æ¢ä¸ºå­—ç¬¦ä¸²çš„å‡½æ•°ï¼ˆè¿™é‡Œå¯¼å…¥ä½†ä»£ç æœªæ˜ç¡®ä½¿ç”¨ï¼‰
import subprocess
# subprocess ç”¨äºæ‰§è¡Œå¤–éƒ¨ shell å‘½ä»¤ï¼ˆè„šæœ¬æœ«å°¾ç”¨äºåˆ é™¤ä¸´æ—¶æ–‡ä»¶ç­‰ï¼‰

# ---------- read_anno å‡½æ•°ï¼šè¯»å–æ³¨é‡Šè¡¨ï¼ˆUCSC-all è¡¨æ ¼æ ·å¼ï¼‰ ----------
def read_anno(fn):
    output = defaultdict(dict)
    # output: ä¸€ä¸ª defaultdictï¼Œå®ƒçš„æ¯ä¸ªé”®ï¼ˆtranscript idï¼‰å¯¹åº”ä¸€ä¸ª dictï¼ˆä¿å­˜ chr, dir, exons ç­‰ï¼‰
    with open(fn, 'r') as input:
        line = input.readline()
        while (line):
            line = line.strip().split("\t")
            # æŠŠä¸€è¡ŒæŒ‰åˆ¶è¡¨ç¬¦åˆ†å‰²æˆåˆ—åˆ—è¡¨
            trans_id = line[1]
            # è½¬å½•æœ¬ IDï¼ˆUCSC all è¡¨ç¬¬äºŒåˆ—ï¼‰
            chr = line[2]
            # æŸ“è‰²ä½“åï¼ˆç¬¬ä¸‰åˆ—ï¼‰
            dir = line[3]
            # è½¬å½•æ–¹å‘ï¼ˆç¬¬å››åˆ—ï¼Œ"+" æˆ– "-"ï¼‰
            exonCounts = int(line[8])
            # å¤–æ˜¾å­æ•°é‡ï¼ˆç¬¬ä¹åˆ—ï¼‰ï¼Œè½¬æˆæ•´æ•°
            exonStarts = map(lambda x: int(x), line[9].split(",")[:-1])  # Starts are 0-based
            # å¤–æ˜¾å­èµ·å§‹ä½ç‚¹åˆ—è¡¨ï¼ˆç¬¬ååˆ—ï¼‰ï¼ŒUCSC æ ¼å¼ä»¥é€—å·ç»“å°¾ï¼Œæ‰€ä»¥å»æ‰æœ€åä¸€ä¸ªç©ºé¡¹ï¼Œ0-based
            exonEnds = map(lambda x: int(x) - 1, line[10].split(",")[:-1])  # Ends are 1-based
            # å¤–æ˜¾å­ç»“æŸä½ç‚¹åˆ—è¡¨ï¼ˆç¬¬åä¸€åˆ—ï¼‰ï¼ŒUCSC æ˜¯ 1-based ç»“æŸï¼Œè„šæœ¬æŠŠå®ƒå‡ 1 ä»¥å˜ä¸º 0-based
            gene_id = line[12]
            # åŸºå›  IDï¼ˆç¬¬åä¸‰åˆ—ï¼‰
            bins = list(zip(exonStarts, exonEnds))
            # æŠŠ starts å’Œ ends é…å¯¹ä¸º (start,end) çš„åˆ—è¡¨ â€”â€” æ³¨æ„ start ä»¥ 0-baseï¼Œend å·²å‡ 1ï¼ˆå˜æˆ 0-baseï¼‰
            enst_start = 0
            output[trans_id]['dir'] = dir
            output[trans_id]['ensg'] = gene_id
            output[trans_id]['chr'] = chr
            output[trans_id]['introns'] = OrderedDict()
            output[trans_id]['exons'] = OrderedDict()
            # ä¸ºè¯¥è½¬å½•æœ¬åˆå§‹åŒ–ç»“æ„ï¼šä¿å­˜æ–¹å‘ã€åŸºå›  idã€æŸ“è‰²ä½“ï¼Œä»¥åŠ exons/introns çš„ OrderedDict

            last_end = None
            last_start = None
            enst_start = -1  # 0-based
            if dir == "+":
                for start, end in bins:
                    enst_start += 1
                    enst_end = enst_start + end - start
                    output[trans_id]['exons'][(enst_start, enst_end)] = (start, end)
                    enst_start = enst_end
                    last_end = end
            # å¦‚æœæ˜¯æ­£é“¾ï¼šæŒ‰ bins é¡ºåºéå†å¤–æ˜¾å­ï¼ŒæŠŠ transcript åæ ‡æ®µ (enst_start,enst_end) æ˜ å°„åˆ°åŸºå› ç»„åæ ‡ (start,end)
            elif dir == "-":
                bins = bins[::-1]
                for start, end in bins:
                    enst_start += 1
                    enst_end = enst_start + end - start
                    output[trans_id]['exons'][(enst_end, enst_start)] = (start, end)  # Noted that '-' strand is reverse
                    enst_start = enst_end
                    last_start = start
            # å¦‚æœæ˜¯åé“¾ï¼šå…ˆåè½¬ binsï¼ˆå› ä¸ºè½¬å½•æœ¬åæ ‡ä» 5'->3'ï¼‰å¹¶ä¸”æŠŠ transcript ç´¢å¼•æ˜ å°„æ–¹å‘åè¿‡æ¥ï¼ˆé”®å­˜ä¸º (enst_end,enst_start)ï¼‰
            line = input.readline()
    return output
# å‡½æ•°è¿”å›ä¸€ä¸ªä»¥è½¬å½•æœ¬ id ä¸ºé”®çš„å­—å…¸ï¼Œæ¯ä¸ªå€¼åŒ…å« 'dir','ensg','chr','exons'ï¼ˆOrderedDictï¼‰ç­‰ä¿¡æ¯

# ---------- cal å‡½æ•°ï¼šç»Ÿè®¡ CIGAR çš„ä¸åŒæ“ä½œçš„é•¿åº¦ ----------
def cal(cigar):
    c = {0: 0, 1: 0, 2: 0, 3: 0, 4: 0}
    # åˆå§‹åŒ–å­—å…¸ï¼š0=M,1=I,2=D,3=N,4=S ç­‰ï¼Œç»Ÿè®¡æ¯ç§ç±»å‹çš„æ€»é•¿åº¦
    for a, b in cigar:
        c[a] += b
    return c[0], "M" + str(c[0]) + "N" + str(c[3]) + "I" + str(c[1]) + "D" + str(c[2])
    # è¿”å› M æ€»é•¿åº¦ï¼ˆä½œä¸ºæ•°å­—ï¼‰å’Œä¸€ä¸ªç®€çŸ­æè¿°å­—ç¬¦ä¸²ï¼ˆä¾‹å¦‚ "M100N20I0D0"ï¼‰

# ---------- generate_new_cigar å‡½æ•°ï¼šåŸºäºè½¬å½•æœ¬ exon ç»“æ„æŠŠæ—§çš„è½¬å½•æœ¬åæ ‡ CIGAR æ˜ å°„åˆ°åŸºå› ç»„åæ ‡ CIGAR ----------
def generate_new_cigar(all_bins, start, end, old_cigar, trans_dir):
    ''' order: small --> big corrdinate '''
    new_cigar_tmp = []  # no del and insert, with intron
    if trans_dir == "-":
        old_cigar = old_cigar[::-1]
        all_bins = all_bins[::-1]
        start, end = end, start
    # å¦‚æœæ˜¯åå‘è½¬å½•ï¼ˆ'-'ï¼‰ï¼Œåè½¬ old_cigarã€all_binsï¼Œå¹¶äº¤æ¢ start/endï¼ˆå› ä¸ºåœ¨è½¬å½•æœ¬åæ ‡ä¸Šæ–¹å‘ç›¸åï¼‰
    all_bins_iter = iter(all_bins)
    while (1):
        try:
            x, y = next(all_bins_iter)
            if x <= start <= y < end:
                new_cigar_tmp.append([0, y - start + 1])
                exon_edge = y
            elif x <= start <= end <= y:
                new_cigar_tmp.append([0, end - start + 1])
                break
            elif start < x <= y < end:
                if x - exon_edge - 1 > 0:
                    new_cigar_tmp.append([3, x - exon_edge - 1])
                new_cigar_tmp.append([0, y - x + 1])
                exon_edge = y
            elif start < x <= end <= y:
                if x - exon_edge - 1 > 0:
                    new_cigar_tmp.append([3, x - exon_edge - 1])
                new_cigar_tmp.append([0, end - x + 1])
                break
        except StopIteration:
            sys.exit()
    # ä¸Šé¢è¿™æ®µï¼šéå†æ‰€æœ‰ exon åŒºé—´ all_binsï¼ˆåŸºå› ç»„åæ ‡ï¼‰ï¼ŒæŠŠç»™å®š transcript çš„ [start,end] æ˜ å°„åˆ°ä¸€ä¸²äº¤æ›¿çš„ exon (0=M) å’Œ intron (3=N) æ®µ
    # new_cigar_tmp å­˜æ”¾ä¸´æ—¶çš„ [type, length]ï¼Œtype 0=M,3=Nï¼ˆæ­¤å¤„ä¸å¤„ç† insert/deleteï¼‰
    # å¦‚æœéå†ç»“æŸå‰æ²¡æœ‰åŒ¹é…åˆ°ç›®æ ‡åŒºé—´ä¼šç›´æ¥ sys.exit()ï¼ˆè„šæœ¬ä¼šé€€å‡ºï¼‰ï¼Œæ³¨æ„è¿™é‡Œæ˜¯æš´åŠ›å¤„ç†ï¼Œè‹¥è¾“å…¥æ³¨é‡Šæˆ–åæ ‡æœ‰è¯¯ä¼šç›´æ¥é€€å‡ºç¨‹åº

    new_cigar_tmp_tmp = []

    new_cigar_tmp_iter = iter(new_cigar_tmp)
    cigar_type, number = next(new_cigar_tmp_iter)
    while (1):
        try:
            cigar_type_1, number_1 = next(new_cigar_tmp_iter)
            if cigar_type == cigar_type_1:
                number = number + number_1
            else:
                new_cigar_tmp_tmp.append([cigar_type, number])
                cigar_type, number = cigar_type_1, number_1
        except StopIteration:
            new_cigar_tmp_tmp.append([cigar_type, number])
            break
    # åˆå¹¶ç›¸é‚»ç›¸åŒç±»å‹çš„æ®µï¼ˆæ¯”å¦‚è¿ç»­ä¸¤ä¸ª M æ®µåˆå¹¶ä¸ºä¸€ä¸ª M æ®µï¼‰

    new_cigar_tmp = new_cigar_tmp_tmp
    new_cigar = []
    # debug
    # old_M, old = cal(old_cigar)
    # NT_M, NT = cal(new_cigar_tmp)

    new_cigar_tmp_iter = iter(new_cigar_tmp)
    block = next(new_cigar_tmp_iter)
    for cigar_type, num in old_cigar:
        try:
            if block[0] == 3:
                new_cigar.append((block[0], block[1]))
                block = next(new_cigar_tmp_iter)
            if cigar_type == 0:  # matched
                if num < block[1]:  # smaller than the original block
                    new_cigar.append((0, num))
                    block[1] = block[1] - num
                elif num == block[1]:  # remove a block
                    new_cigar.append((0, num))
                    block = next(new_cigar_tmp_iter)
                    if block[0] == 3:  # intron
                        new_cigar.append((block[0], block[1]))
                        block = next(new_cigar_tmp_iter)
                else:
                    while num > block[1]:
                        new_cigar.append((0, block[1]))
                        num = num - block[1]
                        block = next(new_cigar_tmp_iter)
                        new_cigar.append((block[0], block[1]))  # intron
                        block = next(new_cigar_tmp_iter)  # until the last exon
                    if num == block[1]:
                        new_cigar.append((0, num))
                        block = next(new_cigar_tmp_iter)
                    elif num < block[1]:
                        block[1] = block[1] - num
                        new_cigar.append((0, num))
                    if block[1] < 0:
                        raise Warning("Block start <0")
            elif cigar_type == 1:  # insert
                new_cigar.append((1, num))
            elif cigar_type == 2:  # del
                if num < block[1]:
                    new_cigar.append((2, num))
                    block[1] = block[1] - num
                elif num == block[1]:
                    new_cigar.append((2, num))
                    block = next(new_cigar_tmp_iter)
                    if block[0] == 3:
                        new_cigar.append((block[0], block[1]))
                        block = next(new_cigar_tmp_iter)
                else:
                    while num > block[1]:
                        new_cigar.append((2, block[1]))
                        num = num - block[1]
                        block = next(new_cigar_tmp_iter)
                        new_cigar.append((block[0], block[1]))  # intron
                        block = next(new_cigar_tmp_iter)  # until the last exon
                    if num == block[1]:
                        new_cigar.append((2, num))
                        block = next(new_cigar_tmp_iter)
                    elif num < block[1]:
                        block[1] = block[1] - num
                        new_cigar.append((2, num))
                    if block[1] < 0:
                        raise Warning("Block start <0")
            elif cigar_type == 3:
                new_cigar.append((3, num))
            elif cigar_type == 4:
                new_cigar.append((4, num))
            elif cigar_type == 5:
                new_cigar.append((5, num))
            elif cigar_type == 6:
                new_cigar.append((6, num))
        except StopIteration:
            continue
    # ä¸Šé¢è¿™ä¸€å¤§æ®µæŠŠåŸå§‹è½¬å½•æœ¬ä¸Šçš„ CIGARï¼ˆold_cigarï¼‰æŒ‰ new_cigar_tmpï¼ˆexon+intron çš„åŒºæ®µï¼‰æ˜ å°„ä¸ºæ–°çš„ CIGAR
    # é€»è¾‘è¦ç‚¹ï¼š
    # - å½“é‡åˆ° Mï¼ˆåŒ¹é…ï¼‰æ—¶ï¼Œä¼šæŠŠ M çš„é•¿åº¦åˆ‡åˆ†åˆ° exon blockï¼ˆ0ï¼‰ä¸Šï¼Œè·¨è¶Š exon é—´ä¼šæ’å…¥ Nï¼ˆ3ï¼‰è¡¨ç¤ºè·¨å†…å«å­
    # - å¯¹äº Iï¼ˆæ’å…¥ï¼‰å’Œ Dï¼ˆç¼ºå¤±ï¼‰å°½é‡ä¿æŒåŸå§‹ä½ç½®å¹¶åˆ‡åˆ†ä»¥åŒ¹é… exon boundary
    # - æ”¯æŒå…¶ä»– CIGAR æ“ä½œç  3/4/5/6ï¼ˆè„šæœ¬ç›´æ¥æŠŠå®ƒä»¬é™„åŠ åˆ°æ–° CIGARï¼‰
    # - è‹¥éå† new_cigar_tmp æ—¶é‡ StopIterationï¼ˆå—ä¸å¤Ÿï¼‰ï¼Œåˆ™è·³è¿‡ï¼ˆcontinueï¼‰â€”â€”é€ æˆéƒ¨åˆ†è¾¹ç•Œæ¡ä»¶ä¸‹å¯èƒ½ä¸¢å¤±æˆ–ä¸å®Œæ•´

    # new_M,new = cal(new_cigar)
    # if new_M != old_M:

    # debug

    return new_cigar
# è¿”å›ä¸€ä¸ªæ–°çš„ CIGARï¼ˆlist of tuplesï¼Œä¸ pysam AlignedSegment.cigar æ ¼å¼ä¸€è‡´ï¼‰

# ---------- map_to_genomeï¼šæŠŠå•æ¡è½¬å½•æœ¬å¯¹é½è®°å½•â€œä¸¾å‡â€åˆ°åŸºå› ç»„åæ ‡å¹¶ç”Ÿæˆæ–°çš„ AlignedSegment ----------
def map_to_genome(segment):
    global UNLIFT, total, lifted, unlifted
    total += 1
    try:
        genome_info = annotation.get(segment.reference_name.split("_AG_converted")[0])
        # ä»è¾“å…¥çš„ segment.reference_nameï¼ˆä¾‹å¦‚è½¬å½•æœ¬ id æˆ–å¸¦åç¼€çš„åå­—ï¼‰ä¸­è·å–åŸå§‹è½¬å½•æœ¬ id
        if options.untreated:
            new_ref_id = header_dict.get(genome_info['chr'])
        else:
            new_ref_id = header_dict.get(genome_info['chr']+"_AG_converted")
        trans_dir = genome_info['dir']
    except TypeError:
        genome_info = None
        new_ref_id = None
    # å¦‚æœåœ¨ annotation ä¸­æ‰¾ä¸åˆ°å¯¹åº”è½¬å½•æœ¬ï¼ˆæˆ– genome_info ä¸ºç©ºï¼‰ï¼Œä¼šè¿›å…¥ except åˆ†æ”¯å¹¶æ ‡è®°ä¸º unlifted

    if genome_info and new_ref_id is not None:
        lifted += 1
        old_start = segment.reference_start  # 0-based
        old_end = segment.reference_end - 1  # 1-based -> å˜æˆ 0-based çš„é—­åŒºé—´ï¼ˆæ‰€ä»¥å‡ 1ï¼‰
        # record old transcript-relative start/end

        new_start = None
        new_end = None
        if trans_dir == "+":
            genome_info_iter = list(genome_info["exons"].items())
        elif trans_dir == "-":
            genome_info_iter = list(genome_info["exons"].items())[::-1]
        list_maxend = []
        for key, values in genome_info_iter:
            list_maxend += [key[0], key[1]]
        len_transcript = max(list_maxend)
        # è®¡ç®— transcript é•¿åº¦ï¼ˆç”¨ enst åæ ‡ç³»çš„æœ€å¤§ç«¯ç‚¹ï¼‰ï¼Œç”¨äºåˆ¤æ–­æ—§çš„ end æ˜¯å¦åœ¨è½¬å½•æœ¬èŒƒå›´å†…

        if old_end <= len_transcript:
            while new_start is None or new_end is None:
                for key, values in genome_info_iter:
                    start, end = key
                    geno_start, geno_end = values
                    if trans_dir == "+":
                        if start <= old_start <= end:
                            new_start = geno_start + old_start - start
                        if start <= old_end <= end:
                            new_end = geno_start + old_end - start
                    elif trans_dir == "-":
                        start, end = end, start
                        if start <= old_end <= end:
                            new_end = geno_start + (end - old_end)
                        if start <= old_start <= end:
                            new_start = geno_start + (end - old_start)
                    else:
                        raise Warning("Transcription direction loss.")
                new_cigar = generate_new_cigar(list(genome_info["exons"].values()), new_start, new_end, segment.cigar,
                                               genome_info['dir'])
                # æ ¹æ®åœ¨åŸºå› ç»„ä¸Šçš„ new_start/new_end ä¸ exons ç»“æ„ç”Ÿæˆæ–°çš„ CIGAR
                qual = segment.query_qualities
                mpq = segment.mapping_quality
                seq = segment.query_sequence

                segment_output = pysam.AlignedSegment()
                segment_output.tags = segment.tags
                # æ–°å»ºä¸€ä¸ª AlignedSegment å¯¹è±¡ï¼Œç”¨ä»¥å†™å…¥è¾“å‡º BAMï¼ˆæ‹·è´ tagsï¼‰
                if trans_dir == "-":
                    new_start, new_end = new_end, new_start
                    qual = qual[::-1]
                    seq = reverse_complement(segment.query_sequence)
                    if segment.is_reverse:
                        segment.is_reverse = False
                        segment.mate_is_reverse = True
                    else:
                        segment.is_reverse = True
                        segment.mate_is_reverse = False
                    segment_output.set_tag("TS", "-")
                # å¦‚æœæ˜¯åå‘é“¾ï¼Œè°ƒæ•´ start/end é¡ºåºï¼Œåè½¬è´¨é‡ï¼Œå¹¶è®©åºåˆ—å–äº’è¡¥åå‘ï¼›åŒæ—¶è°ƒæ•´ flagï¼ˆis_reverse, mate_is_reverseï¼‰
                # å¹¶æ‰“ä¸Šè‡ªå®šä¹‰ tag "TS" è¡¨ç¤ºè½¬å½•æ–¹å‘
                # segment_output.set_tag("YG","G2A")
                else:
                    segment_output.set_tag("TS", "+")
                segment_output.query_name = segment.query_name
                segment_output.flag = segment.flag
                segment_output.reference_id = new_ref_id
                segment_output.reference_start = new_start
                segment_output.cigar = new_cigar
                segment_output.query_sequence = seq
                segment_output.query_qualities = qual
                segment_output.mapping_quality = mpq
                # æŠŠä¸»è¦å­—æ®µè®¾ç½®åˆ°æ–°å»ºçš„ AlignedSegment ä¸­
                segment_output.set_tag("GN", genome_info["ensg"])
                segment_output.set_tag("TN", segment.reference_name)
                segment_output.set_tag("TP", segment.reference_start + 1)  # 1-based
                # æ–°å¢æˆ–ä¿ç•™è‹¥å¹² tagï¼š
                # GN: gene idï¼ˆensgï¼‰ï¼ŒTN: åŸå…ˆçš„è½¬å½•æœ¬ reference_nameï¼ŒTP: åŸå§‹è½¬å½•æœ¬ä½ç½®ï¼ˆ1-basedï¼‰

                if segment_output:
                    if options.verify == True:
                        cigar_length = 0
                        for cigar_type, num in segment_output.cigar:
                            if cigar_type == 0 or cigar_type == 1:
                                cigar_length += num
                        if cigar_length != len(segment_output.query_sequence):
                            raise ValueError("Cigar != sequence")
                    return segment_output
                else:
                    segment.reference_name, segment.cigarstring, segment.reference_start, segment.reference_end, genome_info[
                        "chr"], genome_info[
                        "dir"], segment_output.cigarstring, segment_output.reference_start, segment_output.reference_end
        else:
            unlifted += 1
            if options.no_unlift == False:
                UNLIFT.write(segment)
    else:
        unlifted += 1
        if options.no_unlift == False:
            UNLIFT.write(segment)
# å¦‚æœæ— æ³• liftï¼ˆæ— æ³¨é‡Šã€è¶…è¿‡è½¬å½•æœ¬é•¿åº¦ç­‰æƒ…å†µï¼‰ï¼Œè®¡æ•° unlifted å¹¶ï¼ˆå¦‚æœå…è®¸ï¼‰æŠŠåŸå§‹ segment å†™å…¥ä¸€ä¸ª .unlift.bam æ–‡ä»¶

# ---------- ä¸»æµç¨‹ï¼ˆå¦‚æœè„šæœ¬ä½œä¸ºä¸»ç¨‹åºè¿è¡Œï¼‰ ----------
if __name__ == "__main__":
    description = """
	"""
    parser = argparse.ArgumentParser(prog="m5C_mapper", fromfile_prefix_chars='@', description=description,
                                     formatter_class=argparse.RawTextHelpFormatter)
    # argparse å¯¹è±¡ï¼Œç¨‹åºå m5C_mapperï¼Œæ”¯æŒä»æ–‡ä»¶è¯»å–å‚æ•°ï¼ˆ@fileï¼‰
    # Require
    group_required = parser.add_argument_group("Required")
    group_required.add_argument("--input", "-i", dest="input", required=True, help="input SAM/BAM, default is BAM")
    group_required.add_argument("--output", "-o", dest="output", required=True, help="output SAM/BAM, default is BAM")
    group_required.add_argument("--anno", "-a", dest="anno", required=True, help="UCSC-all table like annotation")
    # å¿…éœ€å‚æ•°ï¼šè¾“å…¥ BAMã€è¾“å‡º BAMã€æ³¨é‡Šæ–‡ä»¶ï¼ˆUCSC all table æ ¼å¼ï¼‰

    # Filter
    group_optional = parser.add_argument_group("Optional")
    group_required.add_argument("--fasta", "-f", dest="fasta", required=False, help="Reference for new header")
    group_required.add_argument("--header", "-H", dest="header", required=False, help="Header file for new header")
    group_optional.add_argument("--no-unlift", dest="no_unlift", default=False, action="store_true",
                                help="Do not report unlifted sequences")
    # é€‰é¡¹ï¼š--no-unlift è¡¨ç¤ºä¸è¦è¾“å‡ºæœª lift çš„ reads åˆ°å•ç‹¬æ–‡ä»¶ï¼ˆé»˜è®¤ä¼šè¾“å‡ºï¼‰
    # group_optional.add_argument("--unlift-to-unmapped",dest="unlift_to_unmapped",default=False,action="store_true",help="Set unlift reads unmapped")
    group_optional.add_argument("--sort", dest="sort", default=False, action="store_true",
                                help="Sort bam (and delete unsort)")
    group_optional.add_argument("--no-del-bam", dest="no_del_bam", default=False, action="store_true",
                                help="Do not del bam file after sorting")
    group_optional.add_argument("--index", dest="index", default=False, action="store_true", help="Index sorted bam")
    group_optional.add_argument("--verify", dest="verify", default=False, action="store_true",
                                help="Check output cigar and length")
    group_other = parser.add_argument_group("Other")
    group_other.add_argument("--version", action="version", version="%(prog)s 1.0")
    parser.add_argument("--untreated", "--untreated", default=False, help="if the input is untreated", action="store_true")
    options = parser.parse_args()
    # è§£æå‘½ä»¤è¡Œå‚æ•°å¹¶èµ‹å€¼ç»™ options å¯¹è±¡

    sys.stderr.write("[%s]Loading annotations...\n" % strftime("%Y-%m-%d %H:%M:%S", time.localtime()))
    annotation = read_anno(options.anno)
    total = 0
    lifted = 0
    unlifted = 0
    # è¯»å–æ³¨é‡Šå¹¶åˆå§‹åŒ–è®¡æ•°å™¨ï¼štotalï¼ˆå¤„ç†çš„ reads æ€»æ•°ï¼‰ï¼Œliftedï¼ˆæˆåŠŸ lift çš„æ•°ï¼‰ï¼Œunliftedï¼ˆå¤±è´¥çš„æ•°ï¼‰

    in_mode = "rb"
    out_mode = "wb"
    # pysam æ–‡ä»¶æ‰“å¼€æ¨¡å¼ï¼šrbï¼ˆè¯»äºŒè¿›åˆ¶ BAMï¼‰ï¼Œwbï¼ˆå†™äºŒè¿›åˆ¶ BAMï¼‰

    sys.stderr.write("[%s]Buidling genome header...\n" % strftime("%Y-%m-%d %H:%M:%S", time.localtime()))
    header = {}
    header['HD'] = {'SO': 'unsorted', 'VN': '1.0'}
    header['SQ'] = []
    if options.fasta:
        filex=options.output.split("trans2Genome.bam")[0]+"ref.dict"
        with open(filex, 'w') as output:
            for seq in SeqIO.parse(options.fasta, 'fasta'):
                length = len(seq.seq)
                header['SQ'].append({"SN": seq.id, "LN": length})
                output.write(seq.id + "\t" + str(length) + "\n")
    else:
        raise Warning("Please provide a reference")
    # æ„å»ºè¾“å‡º BAM çš„ headerï¼ˆSQ å­—æ®µï¼šæŸ“è‰²ä½“åå’Œé•¿åº¦ï¼‰ï¼š
    # - ä»æä¾›çš„ fasta æ–‡ä»¶ä¸­è¯»å–åºåˆ—åå’Œé•¿åº¦å¹¶å†™å…¥ header['SQ']
    # - åŒæ—¶å†™ä¸€ä¸ªç®€å•çš„ .dict æ–‡ä»¶ï¼ˆfilexï¼‰ï¼Œå†…å®¹ä¸º "chr<TAB>length"
    # - å¦‚æœæ²¡æœ‰æä¾› fastaï¼Œä¼šæŠ›å‡ºè­¦å‘Šï¼ˆè„šæœ¬ç»§ç»­å—ï¼Ÿè¿™é‡Œæ˜¯ raise Warningï¼Œä¼šæŠ›å‡º Warning è€Œé Exceptionï¼Œè¡Œä¸ºå¯èƒ½ä¸ç»ˆæ­¢ï¼‰

    if options.no_unlift == False:
        unlift_fn = options.input.replace(".bam", ".unlift.bam")
        with pysam.AlignmentFile(options.input, in_mode) as INPUT, pysam.AlignmentFile(options.output, out_mode,header=header) as OUTPUT,\
                pysam.AlignmentFile(unlift_fn, 'wb', template=INPUT) as UNLIFT:
            header_dict = {}
            n_header = 0
            for header in OUTPUT.header["SQ"]:
                header_dict[header['SN']] = n_header
                n_header += 1
            index=1
            for segment in INPUT:
                segment_output = map_to_genome(segment)
                if segment_output:
                    OUTPUT.write(segment_output)
    else:
        with pysam.AlignmentFile(options.input, in_mode) as INPUT, pysam.AlignmentFile(options.output, out_mode,header=header) as OUTPUT:
            header_dict = {}
            n_header = 0
            for header in OUTPUT.header["SQ"]:
                header_dict[header['SN']] = n_header
                n_header += 1
            for segment in INPUT:
                segment_output = map_to_genome(segment)
                if segment_output:
                    OUTPUT.write(segment_output)
    # æ ¹æ® --no-unlift å‚æ•°åˆ†ä¸¤ç§æ¨¡å¼ï¼š
    # 1) é»˜è®¤ï¼ˆno_unlift Falseï¼‰ï¼šåŒæ—¶æ‰“å¼€ INPUTã€OUTPUTã€UNLIFTï¼ˆUNLIFT ç”¨ template=INPUT ä¿æŒ header/formatï¼‰ï¼Œ
    #    éå† INPUT ä¸­æ¯ä¸ª segmentï¼ˆAlignedSegmentï¼‰ï¼Œè°ƒç”¨ map_to_genome å» transformï¼›è‹¥è¿”å›äº† segment_output åˆ™å†™å…¥ OUTPUTï¼Œ
    #    å¦åˆ™ï¼ˆmap_to_genome å†…éƒ¨ä¼šæŠŠæœª lift çš„å†™å…¥ UNLIFTï¼‰
    # 2) å¦‚æœ --no-unlift Trueï¼šåªæ‰“å¼€ INPUTã€OUTPUTï¼Œä¸å†™ UNLIFT æ–‡ä»¶ï¼ˆæœª lift çš„ reads ç›´æ¥ä¸¢å¼ƒæˆ–å¿½ç•¥ï¼‰

    sys.stderr.write("[%s]Finished.\n  Total: %d\n  Lifted: %d\n  Unlifted: %d\n\n" % (
    strftime("%Y-%m-%d %H:%M:%S", time.localtime()), total, lifted, unlifted))
    # åœ¨æ ‡å‡†é”™è¯¯è¾“å‡ºæ‰“å°å¤„ç†ç»“æœç»Ÿè®¡ï¼ˆæ€»æ•°ã€lift æˆåŠŸæ•°ã€å¤±è´¥æ•°ï¼‰

    if options.sort == True:
        sys.stderr.write("[%s]Sorting bam...\n" % strftime("%Y-%m-%d %H:%M:%S", time.localtime()))
        pysam.sort("-o", options.output.replace(".bam", ".sorted.bam"), options.output)
        if options.index == True:
            sys.stderr.write("[%s]Indexing bam...\n" % strftime("%Y-%m-%d %H:%M:%S", time.localtime()))
            pysam.index(options.output.replace(".bam", ".sorted.bam"))
        if options.no_del_bam == False:
            os.remove(options.output)
    subprocess.call("rm -f "+filex,shell=True)
    # å¯é€‰çš„æ’åº/ç´¢å¼•æ­¥éª¤ï¼š
    # - å¦‚æœ --sortï¼šç”¨ pysam.sort ç”Ÿæˆ .sorted.bam
    # - å¦‚æœ --indexï¼šå¯¹ sorted.bam è¿›è¡Œç´¢å¼•
    # - å¦‚æœ --no-del-bam æœªæŒ‡å®šï¼ˆé»˜è®¤åˆ é™¤ï¼‰ï¼Œåˆ™åˆ é™¤æœªæ’åºçš„ä¸­é—´è¾“å‡º
    # æœ€åç”¨ subprocess åˆ é™¤åœ¨å¼€å§‹æ—¶å†™çš„ä¸´æ—¶ ref.dict æ–‡ä»¶ï¼ˆfilexï¼‰
```

---

# æ€»ç»“ï¼ˆåŠŸèƒ½è¯´æ˜ä¸å…³é”®å®ç°ç‚¹ï¼‰

**æ•´ä½“åŠŸèƒ½**
è¯¥è„šæœ¬çš„ä¸»è¦ç›®çš„å°±æ˜¯æŠŠåŸºäºâ€œè½¬å½•æœ¬/å¤„ç†è¿‡çš„å‚è€ƒï¼ˆä¾‹å¦‚ `_AG_converted`ï¼‰â€ä¸Šå¯¹é½çš„ SAM/BAM è®°å½• **liftï¼ˆä¸¾å‡ï¼‰** å›åŸå§‹åŸºå› ç»„åæ ‡ã€‚å¸¸è§ç”¨ä¾‹åŒ…æ‹¬ï¼šæŠŠå¯¹è½¬å½•æœ¬ï¼ˆæˆ–ç»è¿‡ç¢±åŸºå˜æ¢çš„å‚è€ƒï¼‰ä¸Šçš„æ¯”å¯¹ç»“æœè½¬æ¢å›åŸå§‹åŸºå› ç»„ä¸Šçš„åæ ‡ã€é‡å†™ CIGAR ä»¥åæ˜ è·¨ exon/intron çš„ç»“æ„ã€ä¿ç•™/è°ƒæ•´ FLAG ä¸åºåˆ—æ–¹å‘ä¿¡æ¯ç­‰ã€‚

**ä¸»è¦æ­¥éª¤ï¼ˆé«˜å±‚æ¬¡ï¼‰**

1. è¯»å–æ³¨é‡Šæ–‡ä»¶ï¼ˆ`read_anno`ï¼‰ï¼šæŠŠ UCSC é£æ ¼çš„è¡¨è½¬æˆå†…éƒ¨ç»“æ„ï¼Œä¿å­˜æ¯ä¸ªè½¬å½•æœ¬çš„ exons åˆ—è¡¨æ˜ å°„åˆ°åŸºå› ç»„åæ ‡ã€‚
2. è¯»å–è¾“å…¥ BAMï¼Œéå†æ¯æ¡ AlignedSegmentï¼ˆsegmentï¼‰ã€‚
3. å¯¹æ¯æ¡ segmentï¼Œæ‰¾å‡ºå…¶å¯¹åº”çš„è½¬å½•æœ¬ä¿¡æ¯ï¼ˆannotationï¼‰ï¼›è‹¥æ‰¾åˆ°åˆ™è®¡ç®—è½¬å½•æœ¬åæ ‡åˆ°åŸºå› ç»„åæ ‡çš„ `new_start/new_end`ã€‚
4. åŸºäº exon/ intron è¾¹ç•Œä¸åŸå§‹ CIGARï¼Œè°ƒç”¨ `generate_new_cigar` ç”Ÿæˆæ–°çš„åŸºå› ç»„ä¸Š CIGARï¼ˆåŒ…å« N è¡¨ç¤ºå†…å«å­ï¼‰ã€‚
5. æ„å»ºæ–°çš„ pysam.AlignedSegmentï¼Œå†™å…¥è¾“å‡º BAMï¼ˆåŒæ—¶æ‰“ä¸Šå‡ ä¸ª tagï¼šGNã€TNã€TPã€TSï¼‰ã€‚
6. æœªèƒ½ä¸¾å‡çš„ readsï¼ˆannotation æœªå‘½ä¸­æˆ–åæ ‡è¶…å‡ºï¼‰å†™å…¥ `.unlift.bam`ï¼ˆé™¤éç”¨äº† --no-unliftï¼‰ã€‚
7. å¯é€‰æ’åºã€ç´¢å¼•è¾“å‡º BAM å¹¶æ¸…ç†ä¸­é—´æ–‡ä»¶ã€‚

**å…³é”®å®ç°ç»†èŠ‚ä¸æ½œåœ¨é—®é¢˜**

* æ³¨é‡Šè§£æï¼š`read_anno` å‡å®šæ³¨é‡Šè¡¨æ ¼å¼ä¸¥æ ¼éµå¾ª UCSC allTableï¼ˆå­—æ®µä¸‹æ ‡å›ºå®šï¼‰ï¼Œä¸” exonStarts/exonEnds çš„å¤„ç†å¯¹ end åšäº† `-1`ï¼›è‹¥æ³¨é‡Šæ ¼å¼ä¸é¢„æœŸä¸ç¬¦ä¼šå‡ºé—®é¢˜ã€‚
* æ–¹å‘å¤„ç†ï¼šè„šæœ¬ä¸º `dir == "-"` åšäº†è®¸å¤šåå‘/é€†åºå¤„ç†ï¼ˆåŒ…æ‹¬åå‘ CIGARã€åå‘åºåˆ—äº’è¡¥ã€flip flag ç­‰ï¼‰ï¼Œä½†å¯¹ mate-level çš„å¤„ç†è¾ƒç®€å•ï¼ˆåªæ˜¯åˆ‡æ¢ `mate_is_reverse`ï¼‰ï¼Œéœ€è¦åœ¨é…å¯¹åœºæ™¯ä¸¥æ ¼ç¡®è®¤æ˜¯å¦ç¬¦åˆéœ€æ±‚ã€‚
* CIGAR æ˜ å°„ï¼š`generate_new_cigar` æ˜¯è¯¥è„šæœ¬çš„æ ¸å¿ƒ â€”â€” å®ƒæŠŠ transcript ä¸Šçš„ CIGAR æŒ‰ exon/intron è¾¹ç•Œåˆ‡åˆ†å¹¶æ’å…¥ Nã€‚è¿™ä¸ªé€»è¾‘åŒ…å«å¤§é‡è¾¹ç•Œæ¡ä»¶ï¼ˆä¾‹å¦‚ deletion/insert è·¨è¶Š exon è¾¹ç•Œæ—¶çš„åˆ‡åˆ†ï¼‰ï¼Œéœ€è¦ç”¨çœŸå®æ•°æ®å……åˆ†æµ‹è¯•ï¼Œç‰¹åˆ«æ˜¯å«æœ‰è½¯å‰ªåˆ‡ï¼ˆSï¼‰ï¼Œç¡¬å‰ªåˆ‡ï¼ˆHï¼‰æˆ–å…¶ä»–éå¸¸è§æ“ä½œçš„ readsã€‚
* é”™è¯¯å¤„ç†ï¼š`generate_new_cigar` çš„ `StopIteration -> sys.exit()` ä¼šç›´æ¥é€€å‡ºæ•´ä¸ªè„šæœ¬ï¼Œå¯èƒ½å¯¹æ‰¹é‡å¤„ç†ä¸å‹å¥½ï¼ˆå»ºè®®æ”¹ä¸ºæŠ›å‡ºå¼‚å¸¸å¹¶åœ¨ä¸»å¾ªç¯æ•è·è®°å½•é”™è¯¯ï¼‰ã€‚
* æ€§èƒ½ï¼šé€æ¡è¯»å–å¹¶å¤„ç† pysam AlignedSegment çš„æ–¹å¼æ˜¯å¸¸è§çš„åšæ³•ï¼Œä½†å¦‚æœè¾“å…¥ BAM å¾ˆå¤§ï¼Œå†…å­˜å’Œé€Ÿåº¦è¦æµ‹è¯•ã€‚å¯è€ƒè™‘å¹¶è¡Œæˆ–åˆ†ç‰‡å¤„ç†ã€‚
* header æ„å»ºï¼šé€šè¿‡æä¾› fasta æ„å»º SQ å­—æ®µæ˜¯åˆç†çš„ï¼Œä½†è„šæœ¬è¦æ±‚å¿…é¡»æä¾› fastaï¼ˆå¦åˆ™ raise Warningï¼‰ï¼Œå¹¶ä¸”æŠŠ header ä¸­çš„æŸ“è‰²ä½“ååœ¨åç»­è¿˜ä¼šå¯»æ‰¾ `chr+"_AG_converted"` ç­‰ï¼›è¦ç¡®ä¿ fasta çš„ seq ids ä¸ downstream æœŸæœ›ä¸€è‡´ã€‚
* tag å…¼å®¹æ€§ï¼šè„šæœ¬ä¼šè®¾ç½® GN,TN,TP,TS ç­‰è‡ªå®šä¹‰ tagï¼Œè‹¥åç»­åˆ†ææˆ–è½¯ä»¶ä¸è¯†åˆ«è¿™äº› tag ä¸ä¼šå‡ºé”™ï¼Œä½†è¦æ³¨æ„ tag ç±»å‹/è¯­ä¹‰ï¼ˆTP å­˜æ”¾åŸå§‹è½¬å½•æœ¬ 1-based startï¼‰ã€‚
* verify æ¨¡å¼ï¼šå½“ `--verify` æ‰“å¼€æ—¶ä¼šæ£€æŸ¥ç”Ÿæˆçš„ CIGAR åœ¨ M/I æ“ä½œä¸Šçš„é•¿åº¦æ˜¯å¦ç­‰äºåºåˆ—é•¿åº¦ï¼Œä¸ä¸€è‡´ä¼šæŠ›å¼‚å¸¸ï¼Œè¿™æ˜¯ä¸€ä¸ªæœ‰ç”¨çš„æ ¡éªŒã€‚

RNA åˆ†æï¼ˆç‰¹åˆ«æ˜¯ mâ¶A/mâµC ä¿®é¥°åˆ†æï¼‰é€šå¸¸æœ‰ä¸¤ç§æ¯”å¯¹æ–¹å¼ï¼š

| ç±»å‹    | æ¯”å¯¹åˆ°å“ªé‡Œ          | åæ ‡ç³»ç»Ÿ                  | ç‰¹ç‚¹                 |
| ----- | -------------- | --------------------- | ------------------ |
| è½¬å½•æœ¬æ¯”å¯¹ | æ¯”å¯¹åˆ°è½¬å½•æœ¬åºåˆ—ï¼ˆæ— å†…å«å­ï¼‰ | Transcript coordinate | æ›´å¿«ã€æ›´å‡†ï¼Œä½†æ— æ³•ç›´æ¥å¯¹åº”åŸºå› ç»„ä½ç½® |
| åŸºå› ç»„æ¯”å¯¹ | æ¯”å¯¹åˆ°æ•´æ¡åŸºå› ç»„ï¼ˆæœ‰å†…å«å­ï¼‰ | Genome coordinate     | ç»“æœèƒ½å’Œ GTF æ³¨é‡Šä¸€è‡´      |

ä½†æ˜¯åç»­æ­¥éª¤ï¼ˆæ¯”å¦‚ä¿®é¥°ä½ç‚¹æ³¨é‡Šã€åŸºå› å®šä½ï¼‰éœ€è¦çŸ¥é“**åœ¨åŸºå› ç»„ä¸Šçš„å…·ä½“ä½ç½®ï¼ˆchrã€startã€strandï¼‰**ã€‚
â†’ æ‰€ä»¥è¦â€œæŠŠè½¬å½•æœ¬åæ ‡çš„æ¯”å¯¹ç»“æœï¼Œè½¬æ¢ä¸ºåŸºå› ç»„åæ ‡â€ã€‚

## äºŒã€æ•´ä½“æµç¨‹é€»è¾‘ï¼ˆæ¦‚è§ˆï¼‰

å¯ä»¥ç”¨ä¸€å¥è¯æ€»ç»“ï¼š

> è¯»å–æ³¨é‡Šï¼ˆ.anno æ–‡ä»¶ï¼‰ï¼Œå»ºç«‹è½¬å½•æœ¬åˆ°åŸºå› ç»„çš„åæ ‡æ˜ å°„ï¼›
> ç„¶åéå† BAM æ–‡ä»¶é‡Œçš„æ¯ä¸€æ¡ readï¼ŒæŒ‰è½¬å½•æœ¬æ³¨é‡Šè®¡ç®—å®ƒåœ¨åŸºå› ç»„ä¸Šçš„ä½ç½®ï¼›
> å†é‡æ–°å†™å‡ºä¸€ä¸ªæ–°çš„ BAM æ–‡ä»¶ï¼ˆä½ç½®å˜æˆ genome åæ ‡ï¼‰ã€‚



---

##  ä¸‰ã€

### Step 1. è¯»å–æ³¨é‡Šæ–‡ä»¶ï¼ˆ`--anno`ï¼‰

è°ƒç”¨å‡½æ•°ï¼š

```python
annotation = read_anno(options.anno)
```

ğŸ”¹ è¾“å…¥ï¼šUCSC æ ·å¼çš„æ³¨é‡Šæ–‡ä»¶ï¼ˆåŒ…å«è½¬å½•æœ¬ IDã€æŸ“è‰²ä½“ã€å¤–æ˜¾å­åæ ‡ç­‰ï¼‰
ğŸ”¹ è¾“å‡ºï¼šä¸€ä¸ª Python å­—å…¸ï¼ˆ`annotation`ï¼‰ï¼Œå­˜äº†æ¯ä¸ªè½¬å½•æœ¬çš„å¤–æ˜¾å­ä¿¡æ¯ï¼š

```python
annotation["ENST00000331789"] = {
  "chr": "chr1",
  "dir": "+",
  "exons": OrderedDict([(0,1000):(100000,101000), ...])
}
```

è¿™å°±å‘Šè¯‰è„šæœ¬ï¼š

> æŸä¸ªè½¬å½•æœ¬çš„ç¬¬ä¸€ä¸ªç¢±åŸºåœ¨åŸºå› ç»„ä¸Šå¯¹åº”å“ªé‡Œã€‚

---

###  Step 2. å»ºç«‹æ–°çš„ BAM æ–‡ä»¶å¤´ï¼ˆ`header`ï¼‰

å› ä¸ºä½ è¦è¾“å‡ºçš„æ˜¯ **åŸºå› ç»„åæ ‡ BAM**ï¼Œ
æ‰€ä»¥å®ƒéœ€è¦ç”¨åŸºå› ç»„çš„å‚è€ƒåºåˆ—ï¼ˆfastaï¼‰æ¥åˆ›å»º headerã€‚

è„šæœ¬ä» `--fasta` è¯»å–æ¯æ¡æŸ“è‰²ä½“çš„é•¿åº¦ï¼Œå†™è¿›ï¼š

```python
header['SQ'] = [{"SN": "chr1", "LN": 248956422}, ...]
```

---

###  Step 3. æ‰“å¼€è¾“å…¥ BAM å¹¶åˆ›å»ºè¾“å‡º BAM

```python
pysam.AlignmentFile(options.input, "rb")
pysam.AlignmentFile(options.output, "wb", header=header)
```

è¾“å…¥ BAM æ˜¯â€œè½¬å½•æœ¬åæ ‡â€ï¼›
è¾“å‡º BAM æ˜¯â€œåŸºå› ç»„åæ ‡â€ï¼Œæ‰€ä»¥ header ä¹Ÿå˜äº†ã€‚

---

### Step 4. å¯¹æ¯ä¸€æ¡ readï¼Œæ‰§è¡Œåæ ‡è½¬æ¢

æ ¸å¿ƒå‡½æ•°ï¼š

```python
segment_output = map_to_genome(segment)
```

å®ƒçš„é€»è¾‘æ˜¯ï¼š

1. å…ˆæŸ¥å‡ºè¿™æ¡ read æ¯”å¯¹çš„è½¬å½•æœ¬ IDï¼ˆ`segment.reference_name`ï¼‰ï¼›
2. ä»æ³¨é‡Šä¸­æ‰¾åˆ°å¯¹åº”çš„æŸ“è‰²ä½“ã€æ–¹å‘å’Œå¤–æ˜¾å­åŒºé—´ï¼›
3. æŒ‰ read çš„èµ·æ­¢ä½ç½®ï¼Œåœ¨å¤–æ˜¾å­åæ ‡ä¸­æ‰¾åˆ°å®ƒåœ¨åŸºå› ç»„ä¸Šçš„èµ·æ­¢ï¼›
4. é‡æ–°ç”Ÿæˆæ–°çš„ **CIGAR**ï¼ˆå› ä¸ºç°åœ¨æœ‰å¯èƒ½è·¨å†…å«å­äº†ï¼Œéœ€è¦åŠ  `N`ï¼‰ï¼›
5. åˆ›å»ºä¸€ä¸ªæ–°çš„ `pysam.AlignedSegment()` å¯¹è±¡ï¼Œé‡æ–°è®¾ç½®ï¼š

   * æŸ“è‰²ä½“ï¼ˆchrï¼‰
   * æ–°çš„èµ·å§‹ä½ç½®ï¼ˆ`reference_start`ï¼‰
   * æ–°çš„ `CIGAR`
   * åºåˆ—æ–¹å‘ï¼ˆå¦‚æœæ˜¯è´Ÿé“¾åˆ™åå‘äº’è¡¥ï¼‰

---

### Step 5. å†™å…¥æ–°çš„ BAM æ–‡ä»¶

å¦‚æœè½¬æ¢æˆåŠŸï¼š

```python
OUTPUT.write(segment_output)
```

å¦åˆ™ï¼ˆæ— æ³•è½¬æ¢çš„ readsï¼Œæ¯”å¦‚æ²¡æœ‰å¯¹åº”æ³¨é‡Šï¼‰ï¼š

```python
UNLIFT.write(segment)
```

ä¿å­˜åˆ°ä¸€ä¸ª `.unlift.bam` æ–‡ä»¶ä¸­ã€‚

---

### Step 6. åå¤„ç†ï¼ˆå¯é€‰ï¼‰

æ ¹æ®å‘½ä»¤è¡Œå‚æ•°ï¼Œè¿˜å¯ä»¥ï¼š

* `--sort`ï¼šå¯¹è¾“å‡º BAM æ’åº
* `--index`ï¼šå¯¹æ’åºåçš„ BAM å»ºç«‹ç´¢å¼•
* `--no-del-bam`ï¼šä¿ç•™ä¸­é—´æœªæ’åºçš„ BAM
* åˆ é™¤ä¸´æ—¶æ–‡ä»¶ `.dict`

---

##  å››ã€æœ€ç»ˆè¾“å‡ºçš„æ–‡ä»¶ä»¬

å‡è®¾å‘½ä»¤è¡Œæ˜¯ï¼š

```bash
python trans2Genome.py \
    -i sample.trans.bam \
    -o sample.trans2Genome.bam \
    -a annotation.ucsc \
    -f genome.fa
```

åˆ™è¾“å‡ºï¼š

| æ–‡ä»¶                                 | å†…å®¹          | è¯´æ˜                 |
| ---------------------------------- | ----------- | ------------------ |
| **sample.trans2Genome.bam**        | è½¬æ¢åçš„æ¯”å¯¹ç»“æœ    | æ‰€æœ‰ read çš„åæ ‡æ”¹ä¸ºåŸºå› ç»„åæ ‡ |
| **sample.trans.bam.unlift.bam**    | æ— æ³•è½¬æ¢çš„ reads | æ¯”å¦‚æ‰¾ä¸åˆ°å¯¹åº”è½¬å½•æœ¬çš„        |
| **sample.ref.dict**                | æŸ“è‰²ä½“é•¿åº¦å­—å…¸     | ä¸´æ—¶æ–‡ä»¶               |
| ï¼ˆå¯é€‰ï¼‰sample.trans2Genome.sorted.bam | æ’åºåçš„ç»“æœ      | è‹¥ `--sort` å¯ç”¨      |

---

##  äº”ã€ä¸ºä»€ä¹ˆå¿…é¡»è¿™ä¹ˆåšï¼Ÿ

å› ä¸ºå¾ˆå¤šåˆ†æï¼ˆä¾‹å¦‚ mâµCã€mâ¶Aã€RNA-seq å¯è§†åŒ–ï¼‰éƒ½è¦æ±‚ **ç»“æœä»¥åŸºå› ç»„åæ ‡è¡¨ç¤º**ã€‚

RNA çš„ reads å¾€å¾€æ¯”å¯¹åˆ°â€œè½¬å½•æœ¬â€ï¼ˆæ²¡æœ‰å†…å«å­ï¼‰ï¼Œ
ä½†è¦åœ¨ **åŸºå› ç»„æµè§ˆå™¨ï¼ˆå¦‚ IGVï¼‰** æˆ– **ä¸ GTF æ–‡ä»¶** ä¸€èµ·ç”¨æ—¶ï¼Œå°±å¿…é¡»çŸ¥é“å®ƒä»¬å¯¹åº”çš„æŸ“è‰²ä½“åæ ‡ã€‚

---

## ğŸ§¾ å…­ã€ä¸€å¥è¯æ€»ç»“æ•´æ®µè„šæœ¬

> è¿™æ®µè„šæœ¬çš„ä½œç”¨æ˜¯å°†æ¯”å¯¹åˆ°è½¬å½•æœ¬åæ ‡ç³»çš„ BAM æ–‡ä»¶ï¼Œ
> åˆ©ç”¨æ³¨é‡Šæ–‡ä»¶ä¸­çš„å¤–æ˜¾å­ç»“æ„ä¿¡æ¯ï¼Œ
> é‡æ–°è®¡ç®—æ¯æ¡ read åœ¨åŸºå› ç»„ä¸­çš„çœŸå®åæ ‡ä¸ CIGARï¼Œ
> è¾“å‡ºæ–°çš„åŸºå› ç»„åæ ‡ BAM æ–‡ä»¶ï¼Œä¾›åç»­åˆ†æä½¿ç”¨ã€‚

---


**CIGAR** æ˜¯ BAM/SAM æ–‡ä»¶ä¸­çš„ä¸€ä¸ªéå¸¸æ ¸å¿ƒçš„æ¦‚å¿µï¼Œä»£è¡¨ **æ¯”å¯¹æ“ä½œï¼ˆalignment operationsï¼‰** çš„å‹ç¼©æè¿°ã€‚
å®ƒå‘Šè¯‰æˆ‘ä»¬ â€”â€” **ä¸€æ¡æµ‹åºè¯»æ®µï¼ˆreadï¼‰æ˜¯å¦‚ä½•ä¸å‚è€ƒåŸºå› ç»„ï¼ˆreference genomeï¼‰è¿›è¡Œæ¯”å¯¹çš„**ã€‚

---

## ğŸ§¬ ä¸€ã€CIGAR çš„å…¨ç§°

> **C**ompact **I**diosyncratic **G**apped **A**lignment **R**eport
> å³â€œå‹ç¼©çš„ç‰¹å¾æ€§å¸¦ç¼ºå£çš„æ¯”å¯¹æŠ¥å‘Šâ€ã€‚

---

## ğŸ§© äºŒã€CIGAR å­—ç¬¦ä¸²çš„åŸºæœ¬ç»“æ„

å®ƒæ˜¯ä¸€ä¸ªç”±**æ•°å­—+å­—æ¯**ç»„æˆçš„å­—ç¬¦ä¸²ï¼Œæ¯”å¦‚ï¼š

```
76M
35M2I40M5D10M
10M100N20M
```

æ¯ä¸€æ®µç”±ï¼š

* **æ•°å­—** = è¿ç»­ç¢±åŸºçš„æ•°é‡
* **å­—æ¯** = æ“ä½œç±»å‹

---

## ğŸ“˜ ä¸‰ã€å¸¸è§çš„ CIGAR æ“ä½œç¬¦å«ä¹‰

| æ“ä½œç¬¦   | å«ä¹‰                   | å¯¹ read çš„æ¶ˆè€— | å¯¹å‚è€ƒåŸºå› ç»„çš„æ¶ˆè€— | ä¸¾ä¾‹è¯´æ˜                           |
| ----- | -------------------- | ---------- | --------- | ------------------------------ |
| **M** | åŒ¹é…ï¼ˆmatch æˆ– mismatchï¼‰ | âœ…          | âœ…         | read å’Œ reference éƒ½æœ‰è¿™æ®µåºåˆ—ï¼ˆå¯èƒ½æœ‰é”™é…ï¼‰ |
| **I** | æ’å…¥ï¼ˆinsertionï¼‰        | âœ…          | âŒ         | read æ¯” reference å¤šäº†ä¸€æ®µåºåˆ—        |
| **D** | ç¼ºå¤±ï¼ˆdeletionï¼‰         | âŒ          | âœ…         | read æ¯” reference å°‘äº†ä¸€æ®µåºåˆ—        |
| **N** | è·¨å†…å«å­ï¼ˆintron skipï¼‰    | âŒ          | âœ…         | RNA æ¯”å¯¹æ—¶å¸¸è§ï¼Œè¡¨ç¤ºè·¨è¶Šäº†å†…å«å­             |
| **S** | è½¯å‰ªåˆ‡ï¼ˆsoft clipï¼‰       | âœ…          | âŒ         | read æœ‰éƒ¨åˆ†æœªæ¯”å¯¹ï¼Œä½†ä»ä¿ç•™åœ¨åºåˆ—é‡Œ           |
| **H** | ç¡¬å‰ªåˆ‡ï¼ˆhard clipï¼‰       | âŒ          | âŒ         | read æœ‰éƒ¨åˆ†æœªæ¯”å¯¹ï¼Œä¸”åœ¨åºåˆ—é‡Œè¢«å®Œå…¨åˆ é™¤         |
| **=** | å®Œå…¨åŒ¹é…ï¼ˆsequence matchï¼‰ | âœ…          | âœ…         | read ä¸ reference å®Œå…¨ä¸€è‡´          |
| **X** | é”™é…ï¼ˆmismatchï¼‰         | âœ…          | âœ…         | read ä¸ reference ä¸ä¸€è‡´           |

---

## ğŸ§  å››ã€ä¸¾ä¾‹è§£æ

### ç¤ºä¾‹ 1

```
76M
```

ğŸ‘‰ 76 ä¸ªç¢±åŸºå®Œå…¨è¿ç»­æ¯”å¯¹ä¸Šå‚è€ƒåºåˆ—ã€‚

* readï¼š76bp
* referenceï¼š76bp
  æ²¡æœ‰æ’å…¥ã€ç¼ºå¤±æˆ–é”™é…ä¿¡æ¯ã€‚

---

### ç¤ºä¾‹ 2

```
10M1I5M2D20M
```

ğŸ‘‰ å«æœ‰æ’å…¥ä¸ç¼ºå¤±ï¼š

* 10Mï¼šæ¯”å¯¹ä¸Š10bp
* 1Iï¼šreadå¤šäº†1bpï¼ˆæ’å…¥ï¼‰
* 5Mï¼šå†æ¯”å¯¹ä¸Š5bp
* 2Dï¼šreferenceå¤šäº†2bpï¼ˆç¼ºå¤±ï¼‰
* 20Mï¼šå†æ¯”å¯¹ä¸Š20bp

---

### ç¤ºä¾‹ 3ï¼ˆRNAæ¯”å¯¹å¸¸è§ï¼‰

```
50M1000N50M
```

ğŸ‘‰ è¡¨ç¤ºè·¨è¶Šä¸€ä¸ª **å†…å«å­**ï¼š

* æ¯”å¯¹ä¸Š 50bp çš„å¤–æ˜¾å­1
* è·³è¿‡å‚è€ƒåŸºå› ç»„ä¸Š 1000bpï¼ˆä¸€ä¸ªå†…å«å­ï¼‰
* æ¯”å¯¹ä¸Š 50bp çš„å¤–æ˜¾å­2

è¿™å°±æ˜¯ä¸ºä»€ä¹ˆåœ¨ RNA-seq åˆ†æä¸­ `N` æ“ä½œç¬¦éå¸¸å¸¸è§ã€‚

---

## ğŸ” äº”ã€åœ¨ `pysam` ä¸­çš„ CIGAR è¡¨ç¤º

åœ¨ Python è„šæœ¬é‡Œï¼ˆä¾‹å¦‚ä½ è´´çš„è¿™ä¸ªè„šæœ¬ï¼‰ï¼š

```python
segment.cigar
```

æ˜¯ä¸€ä¸ª **åˆ—è¡¨ï¼ˆlistï¼‰**ï¼Œæ¯ä¸ªå…ƒç´ æ˜¯ `(æ“ä½œç¬¦ç¼–å·, é•¿åº¦)`ï¼Œä¾‹å¦‚ï¼š

```python
[(0, 50), (3, 1000), (0, 50)]
```

è¿™é‡Œï¼š

| æ•°å­— | å«ä¹‰      |
| -- | ------- |
| 0  | Mï¼ˆåŒ¹é…ï¼‰   |
| 1  | Iï¼ˆæ’å…¥ï¼‰   |
| 2  | Dï¼ˆç¼ºå¤±ï¼‰   |
| 3  | Nï¼ˆè·¨å†…å«å­ï¼‰ |
| 4  | Sï¼ˆè½¯å‰ªåˆ‡ï¼‰  |
| 5  | Hï¼ˆç¡¬å‰ªåˆ‡ï¼‰  |

---

## ğŸ§¾ å…­ã€è„šæœ¬ä¸­ä¸ CIGAR ç›¸å…³çš„éƒ¨åˆ†

åœ¨ä½ è´´çš„è„šæœ¬é‡Œï¼Œè¿™å‡ å¤„ç‰¹åˆ«å…³é”®ï¼š

### 1ï¸âƒ£ `def cal(cigar):`

è®¡ç®— CIGAR ä¸­å„ç±»æ“ä½œçš„é•¿åº¦æ€»å’Œï¼ˆç»Ÿè®¡ Mã€Iã€Dã€N çš„æ•°é‡ï¼‰ã€‚

### 2ï¸âƒ£ `def generate_new_cigar(...)`

é‡æ–°ç”Ÿæˆä¸€ä¸ªæ–°çš„ CIGAR å­—ç¬¦ä¸²ï¼Œç”¨äºå°†è½¬å½•æœ¬åæ ‡â€œè½¬åŒ–â€ä¸ºåŸºå› ç»„åæ ‡ï¼ˆtranscript â†’ genomeï¼‰ã€‚
ä¹Ÿå°±æ˜¯è¯´ï¼š
RNA æ¯”å¯¹ç»“æœå¾€å¾€æ˜¯åŸºäºè½¬å½•æœ¬çš„åæ ‡ï¼ˆæ²¡æœ‰å†…å«å­ï¼‰ï¼Œ
è¿™ä¸ªè„šæœ¬è¦è½¬æ¢ä¸ºåŸºå› ç»„åæ ‡ï¼ˆè¦æ’å…¥ Nï¼Œè¡¨ç¤ºè·¨å†…å«å­ï¼‰ã€‚

